### PR TITLE
Integrate Honeybadger for Error monitoring

### PR DESCRIPTION
**Pull request by mandarvaze**

### CODE REVIEW

File: apphelpers/rest/hug.py [hunk 1]
```diff
@@ -10,6 +11,20 @@
 from apphelpers.errors import InvalidSessionError
 from apphelpers.sessions import SessionDBHandler
 
+from honeybadger import Honeybadger
+
+
+# Optional Honeybadger integration. It will kick in
+# only if HONEYBADGER_API_KEY environment variable is set
+hb = None
+API_KEY = os.environ.get('HONEYBADGER_API_KEY', None)
```
Comment (gauravr) [inline apphelpers/rest/hug.py:20]: @mandarvaze we can create `APIFactory::setup_honeybadger(self, api_key):` to pass api key
cc: @shon

File: apphelpers/rest/hug.py [hunk 2]
```diff
@@ -235,7 +264,10 @@ def choose_router(self, f):
...
         print(f'{method_args[0]} [{method.__name__.upper()}] => {f.__module__}:{f.__name__}')
         m = method(*method_args, **method_kw)
-        return m(self.access_wrapper(self.db_tr_wrapper(raise_not_found_on_none(f))))
+        f = self.access_wrapper(
+            notify_honeybadger(
+                self.db_tr_wrapper(
+                    raise_not_found_on_none(f))))
         # NOTE: ^ wrapper ordering is important. access_wrapper needs request which
         # others don't. If access_wrapper comes late in the order it won't be passed
         # request parameter.
```
Comment (gauravr) [inline apphelpers/rest/hug.py:273]: @mandarvaze `return m(f)` is missing

File: apphelpers/rest/hug.py (hunk 1 above)
Comment (mandarvaze) [reply apphelpers/rest/hug.py]: Done. See updated PR.

File: apphelpers/rest/hug.py (hunk 2 above)
Comment (mandarvaze) [reply apphelpers/rest/hug.py:276]: Added.

File: apphelpers/rest/hug.py [hunk 3]
```diff
@@ -127,6 +145,11 @@ def set_context(response, request, context, module):
 
 class APIFactory:
 
+    def setup_honeybadger_monitoring(self, api_key):
+        print("Info: Setting up Honeybadger")
```
Comment (gauravr) [inline apphelpers/rest/hug.py:149]: @mandarvaze I think we can follow here `dbtransaction` wrapper
```
hb = Honeybadger()
hb.configure(api_key=api_key)
self.honeybadger_wrapper = honeybadger_monitoring(hb)
```

File: apphelpers/rest/hug.py [hunk 4]
```diff
@@ -135,6 +158,7 @@ def __init__(self, router, urls_prefix=''):
         self.multi_site_enabled = False
         self.site_identifier = None
         self.urls_prefix = urls_prefix
+        self.hb = None
```
Comment (gauravr) [inline apphelpers/rest/hug.py:161]: @mandarvaze `self.honeybadger_wrapper = phony`

File: apphelpers/rest/hug.py [hunk 5]
```diff
@@ -27,6 +30,21 @@ def wrapper(*ar, **kw):
     return f
 
 
+def honeybadger_wrapper(f, hb):
```
Comment (gauravr) [inline apphelpers/rest/hug.py:33]: @mandarvaze  here we can refer `dbtransaction` wrapper
`def honeybadger_monitoring(hb):`

**Review by gauravr — APPROVED** [review]

**Pull request merged**